emacs_templates/emacs/elisp/lang-javascript.el

```diff
@@ -1,18 +1,50 @@
+;; js2-mode
+;; https://github.com/mooz/js2-mode
 (use-package js2-mode
+  :ensure t
+  :mode
+  ("\\.js$" . js2-mode)
+  ("\\.json$" . js2-jsx-mode)           
   :config
   (custom-set-variables '(js2-strict-inconsistent-return-warning nil))
   (custom-set-variables '(js2-strict-missing-semi-warning nil))
 
   (setq js-indent-level 2)
   (setq js2-indent-level 2)
   (setq js2-basic-offset 2)
-  (add-to-list 'auto-mode-alist '("\\.json$" . js2-mode))
-  (add-to-list 'auto-mode-alist '("\\.js$" . js2-jsx-mode)))
 
-(use-package js2-refactor)
+  ;; tern :- IDE like features for javascript and completion
+  ;; http://ternjs.net/doc/manual.html#emacs
+  (use-package tern
+    :ensure t                                  
+    :config
+    (defun my-js-mode-hook ()
+      "Hook for `js-mode'."
+      (set (make-local-variable 'company-backends)
+           '((company-tern company-files))))
+    (add-hook 'js2-mode-hook 'my-js-mode-hook)
+    (add-hook 'js2-mode-hook 'company-mode))
+
+  (add-hook 'js2-mode-hook 'tern-mode)
+
+  ;; company backend for tern
+  ;; http://ternjs.net/doc/manual.html#emacs
+  (use-package company-tern 
+    :ensure t)
+
+  ;; Run a JavaScript interpreter in an inferior process window
+  ;; https://github.com/redguardtoo/js-comint
+  (use-package js-comint
+    :ensure t             
+    :config
+    (setq inferior-js-program-command "node"))
+
+  ;; js2-refactor :- refactoring options for emacs
+  ;; https://github.com/magnars/js2-refactor.el
+  (use-package js2-refactor :defer t
+    :ensure t                
+    :diminish js2-refactor
+    :bind ("C-c j r" . js2r-add-keybindings-with-prefix)))
```

Comment:
I will at least create a lang-based wiki page and will mention each key. But meanwhile, i do add Python and Go so it would be consistent to add other languages until we move them into their separate pages.